### PR TITLE
Add initial roadmap document

### PR DESCRIPTION
**Pull request by imiric**
This is an RFC of what we had previously discussed. It's best viewed as rendered HTML if you click on "View file".

Let me know if I missed anything, if the timelines don't align with what you had in mind, or if anything else should be changed.

### CODE REVIEW

File: ROADMAP.md
Comment (imiric) [inline ROADMAP.md:1]: It might be good to add a graphical timeline to better visualize this. WDYT?

File: ROADMAP.md
Comment (imiric) [inline ROADMAP.md:101]: This might be too broad/vague to be a reachable goal. WDYT?

Comment (inancgumus) [reply ROADMAP.md:101]: Maybe?
`When we implement all the functions found in other tools, that make sense to include it in the xk6 browser.`

**Comment by inancgumus** [comment]
Wow, @imiric, great work! It made what we're aiming for clearer! Thanks.

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:22]:
```suggestion
These are goals achievable within 3-6 months, tentatively done by Q3 2022.

- **Stabilize the existing functionality and fix any major, high-impact, and blocking bugs.**<br>
  Tool stability is very important, and users shouldn't run into any major showstopping issues.

  *How will we achieve this?*<br>
  By manually testing the application in different scenarios, adding more automated tests, and addressing reports from internal and external users.
  Any issues that have a considerable user impact will be given the highest priority and fixed ASAP.

  *Definition of Done*<br>
  This is difficult to gauge, but at some point, we should stop running into any major issues frequently and should consider this goal as reached.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:41]:
```suggestion
- **Limited Alpha deployment and testing of the extension in k6 Cloud.**<br>

  *How will we achieve this?*<br>
  Soon we will create a small-scale deployment for internal use and specific customers, meant to test the basic integration of the extension in the k6 Cloud.

  *Definition of Done*<br>
  When the extension is usable in k6 Cloud, i.e., scripts can run, test results are shown. Frontend changes are not required at this stage.


- **Update [API documentation](https://k6.io/docs/javascript-api/xk6-browser/).**<br>
  The current documentation is not up-to-date, making using xk6-browser outside of the examples quite difficult.

  *How will we achieve this?*<br>
  We are prioritizing working on documentation in all upcoming cycles.

  *Definition of Done*<br>
  When the documentation reflects the state of the current API.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:57]:
```suggestion
These are goals achievable within 6-12 months, tentatively done by mid-2023.

- **Transition our API to be async/`Promise` based.**<br>
  Currently (April 2022), most of our API is synchronous. This is due to the historical fact that k6 didn't support async behavior because of a missing per VU event loop, which is [now available](https://github.com/grafana/k6/pull/2228).
  Async APIs are important for a browser testing tool since most browser behavior and [CDP](https://chromedevtools.github.io/devtools-protocol/) (the protocol we use to communicate with the browser) is event-based. We need to expose an async API to implement this missing functionality and reach feature parity with tools like Playwright or Puppeteer.

  *How will we achieve this?*<br>
  By gradually transitioning the current API to an async implementation and implementing new features as async.

  *Definition of Done*<br>
  When most of the API can be used asynchronously.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:66]:
```suggestion
- **Beta availability of the extension in k6 Cloud for all users.**<br>

  *How will we achieve this?*<br>
  The deployment should be optimized and the extension thoroughly tested before making it available to all users. Frontend changes should be done at this point, and usage costs (CPU, RAM, storage) and pricing details should be determined, followed by public announcements of the availability. Features such as screen capture, video recording, downloading, and uploading files should be available.

  *Definition of Done*<br>
  When all users are eligible to upgrade to a plan that includes browser testing.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:76]:
```suggestion
- **Increase test code coverage; refactor problematic areas of the codebase; fix "flaky" tests, linter issues, etc.**<br>
  We'll be constantly focusing on this mid- and long-term goal. See issues [#228](https://github.com/grafana/xk6-browser/issues/228), [#241](https://github.com/grafana/xk6-browser/issues/241) and [#58](https://github.com/grafana/xk6-browser/issues/58).

  *How will we achieve this?*<br>
  By gradually adding more unit and functional tests and fixing issues when the opportunity arises. We won't have long stints where we focus exclusively on this goal, but it's something we'll dedicate as much time as possible to.

  *Definition of Done*<br>
  The goal is not to reach 100% code coverage, but realistically speaking, 80 or 90% should be achievable. Our CI test runs should be stable and flake-free, and we should have no linter issues.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:91]:
```suggestion
These are goals achievable after a year, tentatively done in 2024 and beyond.

- **Add support for Firefox and other WebKit-based browsers.**<br>
  Currently (April 2022), our main focus is supporting Chromium-based browsers. We should include other browsers to help users increase their confidence in their applications. The main challenges here will be CDP and the behavior differences between browsers.

  *How will we achieve this?*<br>
  By testing other browsers and fixing issues as they arise.

  *Definition of Done*<br>
  When the latest Firefox and certain WebKit-based browsers are also supported as Chromium.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:101]:
```suggestion
- **Feature parity with Playwright and Puppeteer.**<br>
  Currently, our functionality is limited compared to more mature projects like Playwright and Puppeteer. We plan to expand this gradually and reach or exceed the features offered by other browser automation tools.

  *How will we achieve this?*<br>
  By prioritizing new features to add based on API importance and user feedback. After the short-term stability improvements are made, our main focus will be to add more missing features and close the current feature gap.

  *Definition of Done*<br>
  When all of the functionality found in other tools is implemented in xk6-browser.
```

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:113]:
```suggestion
- **Merge the extension into the k6 tool.**<br>
  k6 and Grafana consider browser testing another strategy under the testing umbrella, along with performance, contract, and other types of testing modern web applications benefit from. As such, the scope of the k6 tool will expand to include functional/E2E testing using browser automation, with the ultimate goal of merging the xk6-browser extension into the k6 tool.

  *How will we achieve this?*<br>
  Once the extension is stable, feature-complete, and widely used in k6 Cloud, we will create a PR to the main k6 repository with the extension code. Then we will deprecate the standalone xk6-browser repository and point users to use the main k6 binary instead.

  This will be a minor breaking change, as scripts will have to import `k6/browser` instead of `k6/x/browser`, and CI jobs will need to change, but the functionality should remain the same.

  *Definition of Done*<br>
  Once the extension has been merged into the k6 tool.
```

Please correct me if I'm wrong, but k6 core is about running and managing VUs, dealing with Goja, async stuff, etc. However, even if they are part of the k6 tool, the modules are, well, modules.

File: ROADMAP.md
Comment (inancgumus) [inline ROADMAP.md:123]:
```suggestion
- **Optimize the k6 Cloud deployment and remove the Beta status.**<br>
  At this point, browser testing should be a well-tested feature in k6 Cloud, and work should be done to optimize the deployment.

  *How will we achieve this?*<br>
  By making changes to optimize test startup times and reduce operating costs. An example could be to use browser pools that are reused between test runs.

  *Definition of Done*<br>
  Difficult to gauge, but whenever we feel comfortable removing the Beta label. Both browser testing and support for it in k6 Cloud should be considered feature complete, and regular maintenance and improvements should follow.
```

**Review by inancgumus — COMMENTED** [review]
Great work! A couple of suggestions. I'll need to review this thoroughly, though.

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md]: Yeah, I changed it to something like that in 89dd838. Thanks.

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md]: In the context of extensions, we usually refer to "k6 core" as the "main k6 repository", and don't make a distinction between "core k6 functionality" and "k6 JS modules".

I changed it slightly in 89dd838. Let me know if it's clearer now.

**Comment by imiric** [comment]
Thanks for the feedback @inancgumus!

I made some of your suggestions in 89dd838, but not all of them. Let me know if you disagree with anything and we can discuss it, but I didn't want to point out all my reasonings as we'd be here all week. :sweat_smile:

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md:1]: We could use the help from our design department for this. :) @sniku Do you think this is a good idea?

File: ROADMAP.md
Comment (inancgumus) [reply ROADMAP.md]: It's crystal clear.

**Review by inancgumus — APPROVED** [review]
LGTM, nice work 👏

File: ROADMAP.md
Comment (sniku) [reply ROADMAP.md:1]: It could be a nice addition, but let's consider this after the PR is merged :smile:

File: ROADMAP.md
Comment (sniku) [inline ROADMAP.md:4]: I think this intro is great. The only thing I would change is to remove the ~undergoing substantial changes~  because it may scare-off potential adopters. Most people read only ~25% of the document so we should start with the description of our ambitions rather than the current state which will change soon.

I would start with this: 
> xk6-browser is an extension that will become part of the k6 core once it reaches its stability goals. The project aims to add real-browser capabilities to k6 to make it more useful to all users—whether they're coming from (...)

File: ROADMAP.md
Comment (sniku) [inline ROADMAP.md:38]: > in all upcoming **development** cycles. 

I don't think our users know what our cycles are :-)

File: ROADMAP.md
Comment (sniku) [inline ROADMAP.md:82]: > tentatively done in 2024 and beyond.

I would not include timelines that are 2 years away. Let's leave the "long-term goals" without dates. Things will change so we can't make predictions that far in the future.

File: ROADMAP.md
Comment (sniku) [inline ROADMAP.md:108]: I think this should be a mid-term goal rather than a long-term. I don't think the extension should be "feature-complete" before it's merged. I think the requirements for merging should be:
1. The API is relatively stable (existing scripts should run in a new release) and
2. The code should be relatively bug-free. (doesn't crash or produce invalid results for 99/100 test runs).

**Review by sniku — CHANGES_REQUESTED** [review]

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md]: Makes sense. I rephrased your suggestion slightly and reworked the intro in 418abbd. Let me know.

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md]: Sure :smile: Changed in d17ee93.

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md]: Indeed. Changed in d17ee93.

File: ROADMAP.md
Comment (imiric) [reply ROADMAP.md]: Sounds good. Let me know if f38387d resolves this.

**Review by sniku — APPROVED** [review]
Looks great. Thanks for making the changes! 

Let's wait for Matt's review before we merge.

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:6]:
```suggestion
We consider browser automation to be an important part of web-application testing, and we have big goals in mind for xk6-browser. In the spirit of transparency, we'd like to share our roadmap for the project with the k6 community.
We hope that users can plan ahead, trusting the k6 and Grafana's commitment to its success.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:8]:
```suggestion
With that in mind, we'll detail some of our short, mid and long-term goals. Most of these will be worked on concurrently, and reaching them will be a gradual process. The timeframes are also not set in stone, but rather serve as tentative targets that the team is aiming for.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:16]:
```suggestion
- **Stabilize the existing functionality and fix any major, high-impact, and blocking bugs.**<br>
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:37]:
```suggestion
  The current documentation is not up-to-date, which makes using xk6-browser outside of the examples quite difficult.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:40]:
```suggestion
  We are prioritizing documentation work in all upcoming development cycles.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:52]:
```suggestion
  Currently (April 2022), most of our API is synchronous. This is due to the historical fact that k6 didn't support async behavior because of a missing per-VU event loop.
[This event loop is now available](https://github.com/grafana/k6/pull/2228).
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:53]:
```suggestion
  Async APIs are important for a browser-testing tool, since most browser behavior and [CDP](https://chromedevtools.github.io/devtools-protocol/) (the protocol we use to communicate with the browser) is event-based. We need to expose an async API to implement this missing functionality and reach feature parity with tools like Playwright or Puppeteer.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:65]:
```suggestion
  The deployment should be optimized and the extension thoroughly tested before making it available to all users. Frontend changes should be done at this point, and usage costs (CPU, RAM, storage) and pricing details should be determined, followed by public announcements of the availability. Features such as screen capture, video recording, downloading, and file uploading should be available.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:72]:
```suggestion
  k6 and Grafana consider browser testing to be another strategy under the testing umbrella, along with performance, contract, and other types of testing that modern web applications benefit from. As such, the scope of the k6 tool will expand to include functional/E2E testing using browser automation, with the ultimate goal of merging the xk6-browser extension into the main k6 repository as a core JS module.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:84]:
```suggestion
  This is both a mid and long-term goal that we'll be constantly focusing on. See issues [#228](https://github.com/grafana/xk6-browser/issues/228), [#241](https://github.com/grafana/xk6-browser/issues/241) and [#58](https://github.com/grafana/xk6-browser/issues/58).
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:99]:
```suggestion
  Currently (April 2022), our main focus is supporting Chromium-based browsers. We should expand support to include other browsers as well. The main challenges here will be around CDP and the behavior differences between browsers.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:105]:
```suggestion
  When the latest Firefox and certain WebKit-based browsers are as well-supported as Chromium is.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:122]:
```suggestion
  By making changes to optimize test-startup times and reduce operating costs. An example could be to use browser pools that are reused between test runs.
```

**Review by MattDodsonEnglish — CHANGES_REQUESTED** [review]
Very nice, well-polished document. All my edits are very minor, finishing touches:
- Making lists parallel
- Using commas before "which", a word that introduces unnecessary, supplementary information in the sentence
- Using hyphens to group nouns collectively functioning as one adjective

File: ROADMAP.md
Comment (inancgumus) [reply ROADMAP.md:84]: ?

```suggestion
  We'll constantly focus on these mid and long-term goals. See issues [#228](https://github.com/grafana/xk6-browser/issues/228), [#241](https://github.com/grafana/xk6-browser/issues/241) and [#58](https://github.com/grafana/xk6-browser/issues/58).
```

File: ROADMAP.md
Comment (inancgumus) [reply ROADMAP.md:37]:
```suggestion
  The current documentation is not up-to-date and makes using xk6-browser outside of the examples quite difficult.
```

File: ROADMAP.md
Comment (MattDodsonEnglish) [inline ROADMAP.md:77]:
```suggestion
  Once the extension and API are relatively stable, and it's well-tested in k6 Cloud, we will create a PR to the main k6 repository with the extension code. Then we will deprecate the standalone xk6-browser repository, and point users to use the main k6 binary instead.
```

**Review by MattDodsonEnglish — APPROVED** [review]

**Pull request merged**
